lib/app_jwt/app_jwt.c

```diff
@@ -0,0 +1,767 @@
+				    size_t message_size, uint8_t *signature, size_t signature_size)
+{
+	psa_status_t status;
+
+	/* Verify the signature of the message */
+	status = psa_verify_message(user_key_id, PSA_ALG_ECDSA(PSA_ALG_SHA_256), message,
+				    message_size, signature, signature_size);
+	if (status != PSA_SUCCESS) {
+		LOG_INF("signature verification failed! (Error: %d)", status);
+		return -1;
+	}
+
+	return 0;
+}
+#endif /* CONFIG_APP_JWT_VERIFY_SIGNATURE */
+
+static int get_random_bytes(uint8_t *output_buffer, const size_t output_length)
+{
+	int err = -ENODATA;
+
+	if ((NULL == output_buffer) || (output_length < 4)) {
+		/* bad parameter */
+		return -EINVAL;
+	}
+
+	psa_status_t psa_status = psa_generate_random(output_buffer, output_length);
+
+	if (PSA_SUCCESS != psa_status) {
+		LOG_ERR("psa_generate_random failed! = %d", psa_status);
+		err = -ENOMEM;
+	} else {
+		err = 0;
+	}
+
+	return err;
+}
+
+static int crypto_finish(void)
+{
+	psa_status_t status;
+
+	/* Purge the key from memory */
+	status = psa_purge_key(IAK_APPLICATION_GEN1);
+	if (status != PSA_SUCCESS) {
+		LOG_ERR("psa_purge_key failed! (Error: %d)", status);
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static char *jwt_header_create(const uint32_t alg, const uint32_t keyid)
+{
+	int err = 0;
+
+	uint8_t pub_key_hash[ECDSA_SHA_256_HASH_SZ];
+
+	char *hdr_str = NULL;
+
+	cJSON *jwt_hdr = cJSON_CreateObject();
+
+	if (!jwt_hdr) {
+		LOG_ERR("cJSON_CreateObject failed!");
+		return NULL;
+	}
+
+	/* Type: format: always "JWT" */
+	if (cJSON_AddStringToObjectCS(jwt_hdr, "typ", "JWT") == NULL) {
+		return NULL;
+	}
+
+	/* Algorithme: format: always "ES256" */
+	if (alg == JWT_ALG_TYPE_ES256) {
+		if (cJSON_AddStringToObjectCS(jwt_hdr, "alg", "ES256") == NULL) {
+			return NULL;
+		}
+	} else {
+		return NULL;
+	}
+
+	/* Keyid: format: sha256 string */
+	/* Get kid: sha256 over public key */
+	size_t olen;
+
+	err = export_public_key_hash(IAK_APPLICATION_GEN1, pub_key_hash, ECDSA_SHA_256_HASH_SZ,
+				     &olen);
+	if (err) {
+		LOG_ERR("Failed to export public key, error: %d", err);
+		return NULL;
+	}
+
+	char sha256_str[ECDSA_SHA_256_SIG_SZ + 1] = {0};
+
+	int32_t printed_bytes = 0;
+
+	for (uint32_t i = 0; i < 8; i++) {
+		printed_bytes +=
+			sprintf((sha256_str + printed_bytes), "%08x",
+				sys_cpu_to_be32((uint32_t)*((uint32_t *)pub_key_hash + i)));
+	}
+	sha256_str[ECDSA_SHA_256_SIG_SZ] = '\0';
+
+	if (cJSON_AddStringToObjectCS(jwt_hdr, "kid", sha256_str) == NULL) {
+		err = -ENOMEM;
+	}
+
+	if (err == 0) {
+		hdr_str = cJSON_PrintUnformatted(jwt_hdr);
+	}
+
+	cJSON_Delete(jwt_hdr);
+	jwt_hdr = NULL;
+
+	return hdr_str;
+}
+
+static char *convert_str_to_b64_url(const char *const str)
+{
+	size_t str_len = strlen(str);
+
+	size_t b64_sz = BASE64_ENCODE_SZ(str_len) + 1;
+
+	char *const b64_out = calloc(b64_sz, 1);
+
+	if (b64_out) {
+		int err = base64_encode(b64_out, b64_sz, &b64_sz, str, strlen(str));
+
+		if (err) {
+			LOG_ERR("base64_encode failed, error: %d", err);
```

Comment:
If encoding fails, it would probably be best to free the output buffer and return `NULL`.